Prelim M150 functionality

JohnTRoth opened the pull request:
Support M150 (Issue #14)

JohnTRoth committed:
Add files via upload

Created handler for M150, supporting R, G/U, B, P, and I parameters

JohnTRoth committed:
Merge pull request #1 from JohnTRoth/JohnTRoth-M150-1

Add files via upload

EricHigdon commented on octoprint_rgb_status/__init__.py:
```diff
@@ -438,10 +439,26 @@ def get_update_information(self, *args, **kwargs):
             }
         }
 
+    def HandleM150(self, comm_instance, phase, cmd, cmd_type, gcode, *args, **kwargs):
```
Use lower_snake_case for method names.

EricHigdon commented on octoprint_rgb_status/basic_effects.py:
```diff
@@ -47,6 +47,16 @@ def solid_color(strip, color, queue, delay=0, iterations=1, reverse=False):
         strip.setPixelColorRGB(p, *color)
     strip.show()
 
+def solid_with_brightness(strip, color, queue, delay=10, iterations=1, reverse=False, brightness=255, index=-1):
```
Let’s just add the brightness and index parameters to the existing solid_color function.

JohnTRoth committed:
Add files via upload

Added fully functional M150, supporting the following:
R - Red value (0-255)
G or U - Green value (0-255)
B - Blue value (0-255)
P - Brightness (0-255)
I - LED Index (1 to # of LED's) When used in combination with Effect 9 (Solid With Brightness), specifying this will control the color/brightness of only the specified LED. Leaving this parameter out (or using a value of -1) will result in controlling all LED's
E - Effect (0-9) The effect to display. Leaving out this parameter defaults to Effect 9 (Solid With Brightness) to maintain compatibility with the official M150 code.

Effects are one of:
0 = Theater Chase
1 = Solid Color
2 = Rainbow Cycle
3 = Rainbow
4 = Color Wipe
5 = Pulse
6 = Knight Rider
7 = Plasma
8 = Theater Chase Rainbow
9 = Color With Brightness

JohnTRoth committed:
Merge pull request #3 from JohnTRoth/JohnTRoth-M150

Add files via upload

EricHigdon commented on octoprint_rgb_status/__init__.py:
```diff
@@ -445,14 +449,33 @@ def HandleM150(self, comm_instance, phase, cmd, cmd_type, gcode, *args, **kwargs
+                    index=parameters['i']
+                )
+            elif hasattr(self, '_queue'):
+                json_str = json.JSONEncoder().encode({'data':{'r': parameters['r'],'g': parameters['g'],'b': parameters['b'],'brightness': parameters['p'],'index': parameters['i']}})
```
Why not use json.dumps() to parallel loads() on line 19 of basic_effects.py?

EricHigdon commented on octoprint_rgb_status/basic_effects.py:
```diff
@@ -42,13 +67,23 @@ def progress_effect(strip, color, queue, delay=0, iterations=1, reverse=False, p
+        for p in range(strip.numPixels()):
+            strip.setPixelColorRGB(p, *color)
+    else:
+        for p in range(strip.numPixels()):
```
This loop is unnecessary. Just set the color of the led at index.

JohnTRoth committed:
Included Delay parameter ('D')

JohnTRoth committed:
Merge pull request #4 from JohnTRoth/JohnTRoth-M150-Delay

Included Delay parameter ('D')

JohnTRoth commented:
Added fully functional M150, supporting the following:
R - Red value (0-255)
G or U - Green value (0-255)
B - Blue value (0-255)
P - Brightness (0-255)
D - Delay
I - LED Index (1 to # of LED's) When used in combination with Effect 9 (Solid With Brightness), specifying this will control the color/brightness of only the specified LED. Leaving this parameter out (or using a value of -1) will result in controlling all LED's
E - Effect (0-9) The effect to display. Leaving out this parameter defaults to Effect 9 (Solid With Brightness) to maintain compatibility with the official M150 code.

Effects are one of:
0 = Theater Chase
1 = Solid Color
2 = Rainbow Cycle
3 = Rainbow
4 = Color Wipe
5 = Pulse
6 = Knight Rider
7 = Plasma
8 = Theater Chase Rainbow
9 = Color With Brightness

No known bugs, but plenty of room for improvement.

EricHigdon commented:
@JohnTRoth Apologies, I left several comments on the code above and just realized that I never finalized my review.
